### PR TITLE
Add TokenType class

### PR DESCRIPTION
**Pull request by connorhu**
The idea of a `TokenType` class was discussed to make it easier to support ORM2 and ORM3 together.
https://github.com/beberlei/DoctrineExtensions/pull/421#discussion_r1478274153

### CODE REVIEW

File: src/Query/TokenType.php [hunk 1]
```diff
@@ -0,0 +1,99 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Doctrine\ORM\Query;
+
+class TokenType
```
Comment (derrabus) [inline src/Query/TokenType.php]:
```suggestion
final class TokenType
```

**Comment by derrabus** [comment]
I suggest we have the old `Lexer` constants point to the new `TokenType` ones and flag the `Lexer` constants as `@deprecated`.

**Comment by connorhu** [comment]
@derrabus Should I deprecate each constant or is one per section enough? I would mark all of them, but I don't know if you have any policy for that.

**Comment by derrabus** [comment]
If we want static analyzers to pick up the deprecation, we need to flag each constant as deprecated.

**Comment by greg0ire** [comment]
I think you should search and replace occurrences of `Lexer::T_` throughout the project (including in docs).

**Review by greg0ire — DISMISSED** [review]

**Comment by greg0ire** [comment]
You will need to update this as well: https://github.com/doctrine/orm/blob/40fbbf4429b0d66517244051237a2bd0616a7a13/psalm-baseline.xml#L2047-L2050

Please take a look at [this guide](https://www.doctrine-project.org/contribute/#dealing-with-checks-and-tools) for more on how to try things locally.

**Review by greg0ire — COMMENTED** [review]
Ah sorry I forgot: please add something in UPGRADE.md regarding the new deprecation.

**Comment by connorhu** [comment]
@greg0ire I made the change you asked for.

**Review by greg0ire — DISMISSED** [review]

File: docs/en/reference/dql-doctrine-query-language.rst [hunk 2]
```diff
@@ -827,12 +827,12 @@ classes have to implement the base class :
...
         {
-            $parser->match(Lexer::T_IDENTIFIER);
-            $parser->match(Lexer::T_OPEN_PARENTHESIS);
+            $parser->match(TokenType::T_IDENTIFIER);
+            $parser->match(TokenType::T_OPEN_PARENTHESIS);
 
             $this->simpleArithmeticExpression = $parser->SimpleArithmeticExpression();
 
-            $parser->match(Lexer::T_CLOSE_PARENTHESIS);
+            $parser->match(TokenType::T_CLOSE_PARENTHESIS);
```
Comment (derrabus) [inline docs/en/reference/dql-doctrine-query-language.rst:835]: Please fix the import in line 815.

File: src/Query/Lexer.php [hunk 3]
```diff
@@ -21,95 +21,249 @@
...
 class Lexer extends AbstractLexer
 {
     // All tokens that are not valid identifiers must be < 100
-    public const T_NONE              = 1;
-    public const T_INTEGER           = 2;
-    public const T_STRING            = 3;
-    public const T_INPUT_PARAMETER   = 4;
-    public const T_FLOAT             = 5;
+    /** @deprecated use Doctrine\ORM\Query\TokenType */
+    public const T_NONE = 1;
```
Comment (derrabus) [inline src/Query/Lexer.php]:
```suggestion
    /** @deprecated use {@see TokenType::T_NONE} */
    public const T_NONE = TokenType::T_NONE;
```

**Review by derrabus — CHANGES_REQUESTED** [review]

**Comment by connorhu** [comment]
@derrabus I did this deprecation notation form based on the phpdoc in the ORMException file, which means it is in "wrong" format too. Should I fix it in a separate PR that file to use `@see`?

https://github.com/doctrine/orm/blob/40fbbf4429b0d66517244051237a2bd0616a7a13/src/ORMException.php#L18

File: src/Query/Lexer.php [hunk 4]
```diff
@@ -21,95 +21,249 @@
...
+
+    /** @deprecated use {@see TokenType::T_CLOSE_CURLY_BRACE} */
+    public const T_CLOSE_CURLY_BRACE = TokenType::T_CLOSE_CURLY_BRACE;
 
     // All tokens that are identifiers or keywords that could be considered as identifiers should be >= 100
-    /** @deprecated No Replacement planned. */
-    public const T_ALIASED_NAME         = 100;
-    public const T_FULLY_QUALIFIED_NAME = 101;
-    public const T_IDENTIFIER           = 102;
+    /** @deprecated use {@see TokenType::T_ALIASED_NAME} */
```
Comment (derrabus) [inline src/Query/Lexer.php]:
```suggestion
    /** @deprecated No Replacement planned. */
```

**Review by derrabus — APPROVED** [review]

**Comment by connorhu** [comment]
Thank you for your help in the creation of the PR and thank you for merging it.

**Comment by derrabus** [comment]
We missed a couple of references: #11234

**Comment by connorhu** [comment]
It didn't occur to me to run phpstan with `phpstan-deprecation-rules` rules. That would have caught these mistakes.

**Comment by derrabus** [comment]
We have Psalm for that, but we apparently allow deprecated constants to be used inside the Lexer class. 🙈

File: src/Query/TokenType.php [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+    public const T_LOWER_THAN        = 13;
+    public const T_MINUS             = 14;
+    public const T_MULTIPLY          = 15;
+    public const T_NEGATE            = 16;
+    public const T_PLUS              = 17;
+    public const T_OPEN_CURLY_BRACE  = 18;
+    public const T_CLOSE_CURLY_BRACE = 19;
+
+    // All tokens that are identifiers or keywords that could be considered as identifiers should be >= 100
+    /** @deprecated No Replacement planned. */
```
Comment (stof) [inline src/Query/TokenType.php:31]: why adding a deprecated one in TokenType ?

Comment (derrabus) [reply src/Query/TokenType.php:31]: Consistency, mainly. We've moved all token constants over to the new class.

**Pull request merged**